openscm/adapters/dice.py

```diff
@@ -0,0 +1,388 @@
+
+        self._values.force = np.empty(self._timestep_count)
+        self._values.force[0] = (
+            self._values.fco22x * log2(self._values.M_atm0 / self._values.M_atm_eq)
+            + self._values.fex0
+        )
+
+    def _shutdown(self) -> None:
+        pass
+
+    def _run(self) -> None:
+        """
+        Run the model over the full time range.
+        """
+        for _ in range(self._timestep_count - 1):
+            self._calc_step()
+        self._update_output()
+
+    def _step(self) -> None:
+        """
+        Do a single time step.
+        """
+        self._calc_step()
+        self._update_output()
+
+    def _update_output(self) -> None:
+        """
+        Set output data from values.
+        """
+        self._views.M_atm.set(self._values.M_atm)
+        self._views.M_l.set(self._values.M_l)
+        self._views.M_u.set(self._values.M_u)
+        self._views.T_atm.set(self._values.T_atm)
+        self._views.T_ocean.set(self._values.T_ocean)
+        self._views.force.set(self._values.force)
+
+    def _calc_step(self) -> None:
+        """
+        Calculate a single time step.
+        """
+        self._timestep += 1
+        self._current_time += YEAR
+
+        # Atmospheric pool size
+        self._values.M_atm[self._timestep] = max(
+            self._values.M_atm_lower,
+            self._values.M_atm[self._timestep - 1] * self._values.b11
+            + self._values.M_u[self._timestep - 1] * self._values.b21
+            + self._values.E[self._timestep - 1] * self._period_length / YEAR / 3.666,
+        )
+
+        # Lower ocean pool size (rel. to 1750)
+        self._values.M_l[self._timestep] = max(
+            self._values.M_l_lower,
+            self._values.M_l[self._timestep - 1] * self._values.b33
+            + self._values.M_u[self._timestep - 1] * self._values.b23,
+        )
+
+        # Shallow ocean pool size (rel. to 1750)
+        self._values.M_u[self._timestep] = max(
+            self._values.M_u_lower,
+            self._values.M_atm[self._timestep - 1] * self._values.b12
+            + self._values.M_u[self._timestep - 1] * self._values.b22
+            + self._values.M_l[self._timestep - 1] * self._values.b32,
+        )
+
+        # Increase in temperature of lower oceans (rel. to 1900)
+        self._values.T_ocean[self._timestep] = max(
+            self._values.T_ocean_lower,
+            min(
+                self._values.T_ocean_upper,
+                self._values.T_ocean[self._timestep - 1]
+                + self._values.c4
+                * (
+                    self._values.T_atm[self._timestep - 1]
+                    - self._values.T_ocean[self._timestep - 1]
+                ),
+            ),
+        )
+
+        # Exogenous forcing for other greenhouse gases
+        if self._start_time + self._period_length * self._timestep > self._year2100:
+            forcoth = self._values.fex1
+        else:
+            forcoth = (
```

Comment:
is this a linear interpolation or a weighted mean or something?

Reply:
linear interpolation. all code is from the original!